### DISCUSSION TITLE
'Modular' bot in aiogram

### DISCUSSION THREAD

**Discussion by DLumi** [aiogram/aiogram#522] [Q&A] [answered]
I'd like to store different message functions in different files, and then call them once DP says it's time to do so. Is that an option? Because right now I have a bunch of spaghetti code which is a mixture of message/callback handlers in one 600+ lines file. If I want to add any functionality in the future, it would just visually clog up the bot.py file even more.

**Comment by DLumi** [comment, marked as answer]
I've found what I was looking for, though, it surprises me how little info there's on the framework online.
If anyone wonders how to do it, here is the topic that helped me: https://github.com/aiogram/aiogram/issues/378

Though, I'll try to give my own explanation on the topic.
You should have your entry point, i.e. your main file (`__main__.py` in my case), and the modules that you want (e.g. `message_handler.py`, `callback_handler.py`). You might want to put the message and the callback handlers separately. 
My folder structure looks like this:
```
Bot
-handlers
--callback
--message
__main__.py
```

Here's your `__main__`:
```python
from handlers.message import message_handler
from handlers.callback import callback_handler

...

if __name__ == '__main__':
    message_handler.setup(dp)
    callback_handler.setup(dp)
    executor.start_polling(dp, skip_updates=True)
```

Here's `message_handler.py`:
```python
async def hi(message: aiogram.types.Message, state: aiogram.dispatcher.FSMContext):
    await message.answer('Hi~!')

def setup(dp: Dispatcher):
    dp.register_message_handler(hi, content_types=['text'], state='*')
```

Here's `callback_handler.py`:
```python
async def some_callback(call: aiogram.types.CallbackQuery, state: aiogram.dispatcher.FSMContext):
    await bot.answer_callback_query(call.id)

    await message.answer('Hi~! But from an inline button')

def setup(dp: Dispatcher):
    dp.register_callback_query_handler(some_callback, lambda call: call.data == 'some_callback_data')
```

Once you distribute them all like that, you should be good to go. Just make sure you don't import anything from `__main__.py`, cause otherwise, you'll get an **ImportError**. If you need something from  `__main__.py`, then just put it into another file.

**Reply by IvanShvedov** [reply]
in callback_handler u use "bot", how i can import it?

**Reply by DLumi** [reply]
> in callback_handler u use "bot", how i can import it?

Just assign it to some variable like so: `bot = aiogram.Bot(TOKEN)`
Then put in whatever file you want, and now you can easily import it into your callback files. Just make sure you don't define it in `__main__.py` (or whatever your main file is) cause then you'll get a circular imports error.

**Reply by polo7** [reply]
Hey there, thank you for this. I am interested in breaking my bot apart into some modules too. This method seems little bit complicated to to me. But I don't have any other though :).  What do you think about this approach now after time has passed? Did you use it in your projects?

**Reply by merura** [reply]
> you can easily import

That is really bad solution in context of modular bot. It would be better to
`bot = Bot.get_current() `
or adding kwarg to handler `bot: Bot`
aiogram directly passes bot instance to handlers.

**Comment by bandirom** [comment]
#### Best solution

```python
# file_1.py
from aiogram import Router

router = Router()

@router.message(CommandStart())
async def command_start_handler(message: Message) -> None:
    pass
```

```python 

# main.py
from file1 import router

dp = Dispatcher()

dp.include_router(tg_router)
dp.start_polling(bot)

```

**Reply by blurdn** [reply]
How can you pass bot instanse in other files using this?

**Reply by merura** [reply]
@Bluryxz Just add the argument to handler function 
```python
@router.message(CommandStart())
async def command_start_handler(message: Message, bot: Bot) -> None:
    pass
```
> Notice that the first argument have to be an update (Message, CallbackQuery, etc) object
